Update to v2 of api

DanWilliamsAu opened the pull request:
This tool currently users v1 of the Smartling translation API.  This PR updates the Smartling CLI tool to use [v2 of the Smartling API](https://developer.smartling.com/v2.0/reference) 

# Approach

The approach that I took with this upgrade was to 'get it working' on the new API.  

It currently does not use any of the new features of the API.  There are some interesting features, such as the [strings API](https://developer.smartling.com/v2.0/reference#create-strings) and the [jobs API](https://developer.smartling.com/v2.0/reference#job) and the [get translations endpoint](https://developer.smartling.com/v2.0/reference#post_projects-projectid-locales-localeid-file-get-translations)

Once we have this merged, it would be good to spend some time thinking about how these new features might help with our translations workflow.

# Smartling golang SDK

I have opted to use [Smartling's official golang SDK](https://github.com/Smartling/api-sdk-go) in the CLI tool.  The majority of endpoints we required are implemented there, as well as authenticating and more.  Using the official SDK gets us closer to Smartling itself as well as prevents us from having to re-invent the wheel.

This means this PR removes our implementation of the API v1 Smartling client and library in favour of using the official SDK.  

A follow up item will be to use the official SDK in promotions https://github.com/99designs/promotions/blob/master/Gopkg.toml#L20

Unfortunately, the official SDK is not feature complete - I've had to add some endpoints to the SDK, I plan to get those new features merged into the official SDK (https://github.com/Smartling/api-sdk-go/pull/4)

# Why not just use smartling's official CLI tool?

Smartling has a [cli tool](https://github.com/Smartling/smartling-cli), which is a fork of this library.  Unfortunately that CLI tool is missing some features which means it won't work out of the box for our purposes.  The missing features include:
* a fault tolerant client that can work with smartling's regular errors such as "RESOURCE_LOCKED"
* automatic prefixing with git branch
* caching

DanWilliamsAu committed:
Add official smartling go sdk as a dependency

Plan to have a play around with the sdk, hopefully it supports our needs - this will allow us to remove our own custom smartling client code

DanWilliamsAu committed:
Ls (list) command is working

Commented out a bunch of stuff relating to locales which will need to be fixed

DanWilliamsAu committed:
Hack to get the locales command working

this will need to be tidied up!

DanWilliamsAu committed:
Uploading (put) command is now working

DanWilliamsAu committed:
Rename command is working

Note: rename command was broken before:
* was reading the first arg as both new file name and old file name
* was failing to surface error messages

DanWilliamsAu committed:
Rm command working

Improved in that it now surfaces errors when they occur (previously could fail silently)

DanWilliamsAu committed:
Implement downloading functionality

DanWilliamsAu committed:
Add support to vendored library for extended status endpoint

Also add some helper methods

DanWilliamsAu committed:
Add status functionality to the tool

Using the SDK and v2 of the API

DanWilliamsAu committed:
Tidy up some unused code - things that we now use the SDK for

DanWilliamsAu committed:
Refactor locales command

Put logic inside a locales method on the client.  The locales list endpoint is not in v2, so uses the project detials endpoint instead to achieve the same outcome

DanWilliamsAu committed:
FIXME: last modified command broken

DanWilliamsAu committed:
Read config from smartling.yml file

DanWilliamsAu committed:
Fix some logic errors with stats

DanWilliamsAu committed:
Integrate projects push and pull commands with smartling SDK

DanWilliamsAu committed:
Use sdk for API errors

DanWilliamsAu committed:
Implement the last modified api call

DanWilliamsAu committed:
remove a bunch of code that is no longer needed

DanWilliamsAu committed:
Update dependencies - remove go-querystring

The smartling SDK handles queries, so we don't need this dependency any more

DanWilliamsAu committed:
Progress of updating config reading

DanWilliamsAu committed:
Bring back the status.go file, turns out it is actually needed

DanWilliamsAu committed:
Remove timeout flag option

The responsibility of the client is handled by the smartling SDK.  Leaving this timeout flag out because it isn't currently supported by the SDK.  We can look at returning support for it, if it turns out that we need it.

DanWilliamsAu committed:
Remove the list conditions option

The conditions were supported in API v1, they are no longer relevant to API v2

DanWilliamsAu committed:
No longer refer to the smartling sdk as smartlingNew

DanWilliamsAu committed:
Remove long option from ls command

The list API no longer returns information about string count, so doesn't really make sense any more.  Could implement with the status endpoint, but leaving it out for now.

DanWilliamsAu committed:
TIdy up config reading to include userID

DanWilliamsAu committed:
Update README

sinamt commented on cli/smartling/main.go:
```diff
@@ -44,31 +43,27 @@ var cmdBefore = func(c *cli.Context) error {
 	}
-
-	var sc *smartling.Client
-	if c.Bool("sandbox") {
```
We don't need sandbox support for anything?

DanWilliamsAu replied:
Sandbox was an api v1 thing. There is no equivalent sandbox mode for API v2.

sinamt commented on cli/smartling/main.go:
```diff
@@ -44,31 +43,27 @@ var cmdBefore = func(c *cli.Context) error {
 	}
 
-	if timeout != 0 {
-		sc.SetHttpTimeout(time.Duration(timeout) * time.Second)
```
Is a timeout still supported somewhere?

DanWilliamsAu replied:
Good question - i had removed that, but there was no reason to do so. I've added the timeout setting functionality back.

sinamt reviewed (approved):
Nice work, looks pretty solid to me.

DanWilliamsAu committed:
Add configurable timeout for http client

DanWilliamsAu committed:
Fix resource locked retry error

Was expecting a smartling api error, but it was actually just a reular old error

mtibben commented on cli/smartling/api.go:
```diff
@@ -166,17 +141,16 @@ var PutCommand = cli.Command{
 		}
 
 		localpath := c.Args().Get(0)
-		remotepath := c.Args().Get(1)
```
Hmm how are we getting away with not having a remote path? Seems weird to me not specifying the remote path, how do we know where the file is going?

DanWilliamsAu replied:
The API v2 does not allow you to specify a remote path: https://developer.smartling.com/v2.0/reference#post_projects-projectid-file

You just provide the fileUri (in this case the localpath) and you can then use that to request the file back.

mtibben replied:
looks like two different fields to me?
```
file = The file contents to upload
fileUri = Value that uniquely identifies the uploaded file
```

DanWilliamsAu replied:
`file` is the actual _contents_ of the file, it does not having anything to do with the path of the file or identifying the file.  You can see how it is used here:  https://github.com/99designs/smartling/blob/master/api.go#L171

mtibben replied:
Yeah exactly, the local file is just the content, so why remove the remotepath part? I'm not sure the localpath = remotepath is a good assumption. Some examples of other unix cli tools always use `<source> <target>` arguments

```
$ cp
usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file
       cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory
```

```
$ scp
usage: scp [-346BCpqrv] [-c cipher] [-F ssh_config] [-i identity_file]
           [-l limit] [-o ssh_option] [-P port] [-S program]
           [[user@]host1:]file1 ... [[user@]host2:]file2
```

```
sftp put [-afPpr] local-path [remote-path]
```

DanWilliamsAu replied:
ahhh ok I see what you mean now, the penny has finally dropped! 

I'll patch that up.
